### PR TITLE
feat: refactor match_order_internal

### PR DESCRIPTION
**Pull request by partylikeits1983**
Opening an initial PR to tackle #74. 

### Reasoning:

To do this, fees need to be debited from open orders in `match_order_internal` not in `open_order_internal`, however, the first step should be to clean up the `match_order_internal` function. 

### Changes Made: 
Currently the `match_order_internal` is not composable, and has several nested if / else code blocks which makes it hard to read and audit. 

In this PR, I refactor the `match_order_internal` function to:
1) Remove nested if/else code blocks
2) Create `update_order_storage_and_reward` to handle order updates and matcher rewards
3) Create `emit_match_events` to improve readability of the `match_order_internal` function.

### CODE REVIEW

**Comment by partylikeits1983** [comment]
Fixed #74. 

Added tests to withdraw protocol fees requested by @dcpp 

Overall, improved the readability of `match_order_internal` and removed the fee debiting from user's positions.

File: market-contract/src/main.sw [hunk 1]
```diff
@@ -855,112 +853,49 @@ fn match_order_internal(
...
-                .asset_type == AssetType::Base
-                || s_order
-                    .price < b_order
-                    .price
-                && order0
-                    .asset_type == AssetType::Quote
-        {
+
+        // Checking if the prices align for a possible match
+        if (s_order.price > b_order.price && order0.asset_type == AssetType::Base) || (s_order.price < b_order.price && order0.asset_type == AssetType::Quote) {
```
Comment (dcpp) [inline market-contract/src/main.sw:890]: change order0 -> s_order

File: market-contract/src/main.sw [hunk 2]
```diff
@@ -971,103 +906,176 @@ fn match_order_internal(
...
+    order0: Order,
+    amount0: u64,
+    id1: b256,
+    order1: Order,
+    amount1: u64,
+    matcher: Identity,
+    match_price: u64,
+) {
+    // Emit events for the first order
+    log_order_change_info(
```
Comment (dcpp) [inline market-contract/src/main.sw:1012]: Ask Alex should we remove log_order_change_info. IT was created for indexer debug 3 months ago

File: market-contract/tests/functions/core/withdraw_protocol_fee.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        let accumulated_protocol_fee = contract.total_protocol_fee().await.unwrap().value;
+        assert_eq!(accumulated_protocol_fee, 0);
+
+        contract.match_order_pair(id0, id1).await?;
+
+        let accumulated_protocol_fee = contract.total_protocol_fee().await.unwrap().value;
+        assert!(accumulated_protocol_fee != 0);
+
+        let withdraw_tx = contract.withdraw_protocol_fee(user0.identity()).await;
```
Comment (dcpp) [inline market-contract/tests/functions/core/withdraw_protocol_fee.rs:130]: check that fuel asset was actually withdrawn. check user and contract balance. and total_protocol_fee should be 0 after

Comment (partylikeits1983) [reply market-contract/tests/functions/core/withdraw_protocol_fee.rs:131]: Fixed

File: market-contract/src/main.sw (hunk 2 above)
Comment (partylikeits1983) [reply market-contract/src/main.sw:1012]: Will do.

File: market-contract/src/main.sw (hunk 1 above)
Comment (partylikeits1983) [reply market-contract/src/main.sw]: Fixed.

**Pull request closed without merging**